services/analyzer/mongo-embedded-analyzer.js

```diff
@@ -0,0 +1,231 @@
+      && Array.isArray(structure[currentKey]) !== Array.isArray(type)) {
+      // eslint-disable-next-line no-param-reassign
+      structure[currentKey] = 'mongoose.Mixed';
+    } else {
+      const nestedStructure = [];
+      // eslint-disable-next-line no-param-reassign
+      structure[currentKey] = nestedStructure;
+      type.forEach((valueAnalysed) => {
+        applyType(valueAnalysed, nestedStructure, 0);
+      });
+    }
+  } else if (typeof type === 'object') {
+    if (structure[currentKey] !== undefined) {
+      if (structure[currentKey] === 'mongoose.Mixed' || typeof structure[currentKey] !== 'object') {
+        // Notice: Conflicted type
+        // eslint-disable-next-line no-param-reassign
+        structure[currentKey] = 'mongoose.Mixed';
+      } else if (typeof structure[currentKey] === 'object') {
+        const nestedStructure = structure[currentKey];
+        Object.keys(type).forEach((key) => {
+          applyType(type[key], nestedStructure, key);
+        });
+      } else {
+        const nestedStructure = {};
+        // eslint-disable-next-line no-param-reassign
+        structure[currentKey] = nestedStructure;
+        Object.keys(type).forEach((key) => {
+          applyType(type[key], nestedStructure, key);
+        });
+      }
+    } else {
+      const nestedStructure = {};
+      // eslint-disable-next-line no-param-reassign
+      structure[currentKey] = nestedStructure;
+      Object.keys(type).forEach((key) => {
+        applyType(type[key], nestedStructure, key);
+      });
+    }
+  }
+}
+
+function haveSameEmbeddedType(type1, type2) {
+  return typeof type1 === typeof type2
+    && Array.isArray(type1) === Array.isArray(type2);
+}
+
+function areAnalysesSameEmbeddedType(arrayOfAnalysis) {
+  if (!arrayOfAnalysis && arrayOfAnalysis.length < 2) {
+    return true;
+  }
+
+  for (var i = 0, j = 1; j < arrayOfAnalysis.length; i += 1, j += 1) {
+    if (!haveSameEmbeddedType(arrayOfAnalysis[i], arrayOfAnalysis[j])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function serializeAnalysis(fieldAnalysis) {
+  if (!fieldAnalysis) {
+    return null;
+  }
+
+  const analysis = { type: 'embedded' };
+  analysis.schema = JSON.stringify(fieldAnalysis);
+
+  return analysis;
+}
+
+function deserializeAnalysis(analysis) {
+  if (!analysis) {
+    return null;
+  }
+
+  return JSON.parse(analysis.schema);
+}
+
+function deserializeAnalyses(fieldAnalyses) {
+  if (!fieldAnalyses || fieldAnalyses.length === 0) {
+    return null;
+  }
+
+  const parsedFieldAnalyses = [...fieldAnalyses];
+
+  parsedFieldAnalyses.forEach((analysis, index) => {
+    if (analysis.type === 'embedded') {
+      parsedFieldAnalyses[index] = deserializeAnalysis(analysis);
+    }
+  });
+
+  return parsedFieldAnalyses;
+}
+
+function mergeEmbeddedDetections(keyAnalyses) {
+  if (!areAnalysesSameEmbeddedType(keyAnalyses)) {
+    return 'Object';
+  }
+  const firstAnalysis = keyAnalyses[0];
+  var structure;
```

Comment:
We may use let instead of var I think

Reply:
For some reasons, let doesn't work in map reduce mongo side 😞